src/QuartoGameState.java

```diff
@@ -160,45 +165,107 @@ public void evaluate(QuartoPiece limboPiece) {
+                }
+            }
+        }
+    }
+
+    /**
+     * Tries to find a piece that cannot be placed on the
+     * board to win.
+     *
+     * @return Proposed QuartoPiece
+     */
+    public QuartoPiece getSafePiece() {
+        List<int[]> squaresToTest = this.getTestSquares();
+        boolean safe;
+        for(QuartoPiece limboPiece : this.freePieces) {
+            safe = true;
+            for(int[] square : squaresToTest) {
                 QuartoBoard tmpBoard = new QuartoBoard(this.board);
                 tmpBoard.insertPieceOnBoard(square[0], square[1], limboPiece.getPieceID());
 
-                if(this.hasQuarto(tmpBoard)) {
-                    value = this.isMaxState ? 27 : -27;
-                    break;
+                QuartoGameState state = getRegisteredState(tmpBoard, alpha, beta, this.isMaxState);
+
+                if (state.hasQuarto()) {
+                    safe = false;
+                    continue;
                 }
             }
+
+            if(safe)
+                return limboPiece;
         }
+
+        //No safe moves
+        return this.board.getPiece(this.board.chooseNextPieceNotPlayed());
     }
 
     /**
-     * @return       boolean
+     * Evaluation of free squares to see which are the last
+     * square needed in a possible combination.
+     *
+     * @return A list of nodes that need to be checked
      */
-    // TODO: Get rid of this method
-    public boolean hasQuarto(QuartoBoard tmpBoard)
-    {
-        //loop through rows
-        for(int i = 0; i < tmpBoard.getNumberOfRows(); i++) {
-            if (tmpBoard.checkRow(i)) {
-                return true;
+    private List<int[]> getTestSquares() {
+        if(this.possibleWinSquares == null) {
+            this.possibleWinSquares = new ArrayList<int[]>();
+
+            List<List<int[]>> rows = new ArrayList<List<int[]>>();
+            List<List<int[]>> cols = new ArrayList<List<int[]>>();
+            List<List<int[]>> diags = new ArrayList<List<int[]>>();
+
+            for(int i=0;i<5;i++) {
+                rows.add(new ArrayList<int[]>());
+                cols.add(new ArrayList<int[]>());
+                if(i<2)
+                    diags.add(new ArrayList<int[]>());
             }
-        }
 
-        //loop through columns
-        for(int i = 0; i < tmpBoard.getNumberOfColumns(); i++) {
-            if (tmpBoard.checkColumn(i)) {
-                return true;
+            for (int[] square : this.freeSquares) {
+                rows.get(square[0]).add(square);
+                cols.get(square[1]).add(square);
+                if (square[0] == square[1])
+                    diags.get(0).add(square);
+                else if (square[0] + square[1] == 4)
+                    diags.get(1).add(square);
             }
-        }
 
-        //check Diagonals
-        if (tmpBoard.checkDiagonals()) {
-            return true;
+            for(int i=0;i<5;i++) {
+                if(rows.get(i).size() == 1 &&
```

Comment:
I believe you could simplify this if conditional by making the lower  if's else if.